### PR TITLE
Add eval-only arg

### PR DESCRIPTION
**Pull request by SaulLu**
In this PR I propose to add a `no-train` argument to only execute the evaluation on a checkpoint. This is useful for Alibi because it can be evaluated with different seq lengths. 

This is in my opinion the easiest way to proceed in order to 1) keep exactly the same setting as during the training (and avoid adding divergences) and 2) use exactly the same valid and test split as during the training.

I used these changes to run the [`tr7b-extrapolation-law.slurm`](https://github.com/bigscience-workshop/bigscience/blob/b4a926da7edff66ec11cc01a789baea5fe16a4c0/train/tr7-alibi/tr7b-extrapolation-law.slurm) script introduced in [this PR](https://github.com/bigscience-workshop/bigscience/pull/17).

I've also included the changes necessary in this PR to make it work for Alibi (i.e. deactivate a comparison)

@thomasw21, @TevenLeScao, @ibeltagy, @ofirpress, @slippylolo, @stas00 do you think that is an addition that might be interesting to merge on master?

### CODE REVIEW

**Comment by thomasw21** [comment]
So I usually use the split argument to indicate I want zero training samples. (forgot if this works out of the box, or I had to implement some hacky thing).

Otherwise this would look like an easy to get some evaluation running at least to get loss and PPL.

**Comment by SaulLu** [comment]
> So I usually use the split argument to indicate I want zero training samples. (forgot if this works out of the box, or I had to implement some hacky thing).

Indeed, I remembered that we had discussed this point some time ago! The only difference is that with the current proposal we have the possibility to use exactly the same split as the one used during the training (which is not possible to obtain by playing with the split percentages)

File: megatron/checkpointing.py [hunk 1]
```diff
@@ -61,8 +62,10 @@ def _compare(arg_name, old_arg_name=None):
     _compare('num_layers')
     _compare('hidden_size')
     _compare('num_attention_heads')
-    _compare('max_position_embeddings')
     _compare('position_embedding_type')
+    # with alibi we can change `max_position_embeddings`
+    if args.position_embedding_type != PositionEmbeddingType.alibi:
+        _compare('max_position_embeddings')
```
Comment (stas00) [inline megatron/checkpointing.py:68]: this edit doesn't seem to belong to this PR, right?

**Review by stas00 — APPROVED** [review]
Sounds good to me

But I'd avoid double negative and call the option `--eval-only` ;) 

Though yours is fine as well. Whatever resonates the most.

File: megatron/checkpointing.py (hunk 1 above)
Comment (SaulLu) [reply megatron/checkpointing.py:68]: I included this change because this check becomes blocking when I evaluate a model with Alibi embeddings (which the introduction of `no-train` allows) with a different seq length. 

Do you prefer that I separate it in another PR? :slightly_smiling_face:

Comment (stas00) [reply megatron/checkpointing.py:68]: In general it's a good idea to separate such things, but it's fine for now if you merge it as is.

Comment (SaulLu) [reply megatron/checkpointing.py:68]: > But I'd avoid double negative and call the option --eval-only ;)
That's very wise!

**Comment by SaulLu** [comment]
A small detail, the tests fail at the "Configure AWS credentials" level. It seems to be due to the `Secrets` that it can't access.  Do you have any idea what I should do to solve this problem?

**Comment by thomasw21** [comment]
I think the issue is that it's on your fork, and we're seeing issues with CI of PRs opened on another fork https://github.com/bigscience-workshop/Megatron-DeepSpeed/blob/main/.github/workflows/ci.md#ci-setup
You can always run the tests locally if you are unsure.

**Comment by SaulLu** [comment]
Thanks a lot for the information @thomasw21! Indeed, this seems to be the reason. I have reopened a PR #195 with a branch directly on the repo so that the tests run. And the funny thing is that it activated the tests for this PR too. :exploding_head: 

I will merge this one and close the new one which is obsolete now.

File: megatron/training.py [hunk 2]
```diff
@@ -1046,7 +1046,7 @@ def build_train_valid_test_data_iterators(
                             if test_ds is not None else []
 
         # Flags to know if we need to do training/validation/testing.
-        do_train = (train_dataloader is not None and args.train_iters > 0) and not args.no_train
+        do_train = (train_dataloader is not None and args.train_iters > 0) and not args.eval_only
```
Comment (stas00) [inline megatron/training.py:1074]: not really needing parentheses here as it's: A and B and C, but it's ok.

**Comment by stas00** [comment]
yeah, our CI is not great. We can only run it on non-forked branches. So since you have write access, in the future PRs try to use a local branch.

**Comment by SaulLu** [comment]
> yeah, our CI is not great. We can only run it on non-forked branches. So since you have write access, in the future PRs try to use a local branch. 

Thanks a lot for the info (it's already super nice to have an CI with multiple GPUs!)! I will use a local branch next time

File: megatron/training.py (hunk 2 above)
Comment (SaulLu) [reply megatron/training.py:1074]: Oh yes, that is indeed not very clever of me. I've opened a new PR  #197 to fix this.
